From 656ee00556332659423c5672c2f9398c7e77c2fa Mon Sep 17 00:00:00 2001
From: Hans de Goede <johannes.goede@oss.qualcomm.com>
Date: Sat, 30 May 2026 14:28:51 +0200
Subject: [PATCH] grub-get-kernel-settings: Treate kernel-uki-dtbloader as
 default kernel

kernel-uki-dtbloader is a drop-in replacement for kernel-core, it even
conflicts with kernel-core since it uses identical filenames under /boot.

Before this patch grub-get-kernel-settings handled kernel-uki-dtbloader as
a special kernel variant, causing GRUB_NON_STANDARD_KERNEL=true to get set
which leads to /lib/kernel/install.d/20-grub.install setting
GRUB_UPDATE_DEFAULT_KERNEL=false which results in skipping the
grub2-set-default call later on during kernel-install.

As a result of this users of Fedora 44+ ARM64 live media which uses
kernel-uki-dtbloader would still get the old kernel on reboot after
installing kernel updates (Bug 2463620 - GRUB environment variable
saved_entry not updated when installing UKI kernel on F44 aarch64).

Resolve this by treating kernel-uki-dtbloader as default kernel, just
like how kernel-core is handled.

Note on the next rebase of rhboot/grub this may be merged into
commit c5baa5c1003d ("Add grub-get-kernel-settings and use it in 10_linux")

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=2463620
Fixes: c5baa5c1003d ("Add grub-get-kernel-settings and use it in 10_linux")
Signed-off-by: Hans de Goede <johannes.goede@oss.qualcomm.com>
---
 util/grub-get-kernel-settings.in | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/util/grub-get-kernel-settings.in b/util/grub-get-kernel-settings.in
index f71bc6436..47ea97868 100644
--- a/util/grub-get-kernel-settings.in
+++ b/util/grub-get-kernel-settings.in
@@ -68,7 +68,12 @@ if test -f /etc/sysconfig/kernel ; then
     . /etc/sysconfig/kernel
 fi
 
-GRUB_DEFAULT_KERNEL_TYPE=${DEFAULTKERNEL/-core/}
+# kernel-uki-dtbloader is a drop in replacement for kernel-core, treat it as such
+if [ "$DEFAULTKERNEL" = "kernel-uki-dtbloader"  ]; then
+    GRUB_DEFAULT_KERNEL_TYPE="kernel"
+else
+    GRUB_DEFAULT_KERNEL_TYPE=${DEFAULTKERNEL/-core/}
+fi
 if [ "$GRUB_DEFAULT_KERNEL_TYPE" != "kernel" ]; then
     echo GRUB_NON_STANDARD_KERNEL=true
     echo export GRUB_NON_STANDARD_KERNEL
-- 
2.54.0

